robot-server/robot_server/service/task_runner.py

```diff
@@ -36,15 +34,45 @@ def run(self, func: TaskFunc, **kwargs: Any) -> None:
 
         Arguments:
             func: An async, None-returning function to run in the background.
-            Use kwargs to add arguments if required.
+            Use **kwargs to pass to func.
         """
         func_name = func.__qualname__
 
-        async def _run_async_task() -> None:
+        async def wrapper() -> None:
+            # Add the previous logic back
+            current_task = asyncio.current_task()
             try:
                 await func(**kwargs)
                 log.debug(f"Background task {func_name} succeeded")
             except Exception as e:
                 log.warning(f"Background task {func_name} failed", exc_info=e)
+            finally:
+                self._running_tasks.remove(current_task)
+
+        wrapper_task = asyncio.create_task(wrapper())
+        self._running_tasks.add(wrapper_task)
+
+    async def cancel_all_and_clean_up(self) -> None:
```

Comment:
Docstring suggestion:

```suggestion
    async def cancel_all_and_clean_up(self) -> None:
        """Cancel any ongoing background tasks and wait for them to stop.
        
        Intended to be called just once, when the server shuts down.
        """
```